sycl/source/detail/stream_impl.cpp

```diff
@@ -59,21 +61,33 @@ size_t stream_impl::get_size() const { return BufferSize_; }
 
 size_t stream_impl::get_max_statement_size() const { return MaxStatementSize_; }
 
-void stream_impl::flush() {
-  // Access the stream buffer on the host. This access guarantees that kernel is
-  // executed and buffer contains streamed data.
-  {
-    auto HostAcc = detail::Scheduler::getInstance()
-                       .StreamBuffersPool.find(this)
-                       ->second.Buf.get_access<cl::sycl::access::mode::read>(
-                           range<1>(BufferSize_), id<1>(OffsetSize));
+std::once_flag flag;
```

Comment:
Should this be a static variable?
As far as I understood from the implementation there is a need to launch `std::atexit` only once per application run. `std::call_once` uses a reference to `std::once_flag` from user. Though, this reference is per-`stream_impl` here. Hence we'll call `std::atexit` in per-`stream_impl` manner.

Reply:
Unfortunately, I realized that "atexit" approach to remove the buffers also can't be used, problem is that atexit can be called after all plugins/libraries are unloaded - https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/atexit?view=vs-2019 . sycl buffer destruction calls wait for events underneath, so needs plugins/libraries.

I tried approach to enqueue the second host task with depends_on(event of the print task) but this approach also doesn't work for almost same reasons. 

Leaving destruction of buffers to the scheduler's destructor also doesn't work, because scheduler is a static object and its destructor also called too late (after libraries are unloaded).

For now, I am out of ideas how to guarantee proper resources cleanup.

Reply:
This resource cleaning is complex in SYCL because of https://gitlab.khronos.org/sycl/Specification/-/issues/289

Reply:
Thank you for the link to the discussion!